# Building a Regression Model with Keras: Predicting Concrete Compressive Strength from Multivariate Data

Author: [Arnau Gómez](https://www.arnaugomez.com)

Peer-graded Assignment: Build a Regression Model in Keras

Exercise D: Increase the number of hidden layers (5 marks)

## Initial setup

Import dependencies

In [1]:
# Python version: 3.11.9
%pip install --upgrade pip
%pip install pandas==2.2.3 keras==3.7.0 tensorflow==2.18.0 numpy==2.0.2 scikit-learn==1.6.0

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


Import dependencies

In [2]:
import pandas as pd
import numpy as np
import keras

import warnings
warnings.simplefilter('ignore', FutureWarning)

Import data

In [3]:
filepath='https://cocl.us/concrete_data'
concrete_data = pd.read_csv(filepath)


concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

Split data into predictors and target

In [4]:
predictors = concrete_data.drop(columns=['Strength'])
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [5]:
target = concrete_data['Strength']
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Normalize the data by substracting the mean and dividing by the standard deviation.

In [6]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

Compute number of input features

In [7]:
n_predictors = predictors.shape[1]

## Build the keras model

In [8]:
from keras.models import Sequential
from keras.layers import Dense, Input

# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Input(shape=(n_predictors,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

# build the model
model = regression_model()

# display model summary
model.summary()


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 10)             │            90 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 10)             │           110 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 10)             │           110 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 1)              │            11 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 321 (1.25 KB)

 Trainable params: 321 (1.25 KB)

 Non-trainable params: 0 (0.00 B)

## Train and test the model

Train and test it once:

In [9]:
model.fit(predictors_norm, target, validation_split=0.3, epochs=50, verbose=2)

Epoch 1/50
23/23 - 0s - 19ms/step - loss: 1655.3156 - val_loss: 1192.2191
Epoch 2/50
23/23 - 0s - 2ms/step - loss: 1615.3130 - val_loss: 1160.1373
Epoch 3/50
23/23 - 0s - 2ms/step - loss: 1554.5983 - val_loss: 1109.6897
Epoch 4/50
23/23 - 0s - 2ms/step - loss: 1463.6509 - val_loss: 1036.7717
Epoch 5/50
23/23 - 0s - 1ms/step - loss: 1335.0931 - val_loss: 938.2257
Epoch 6/50
23/23 - 0s - 1ms/step - loss: 1162.8563 - val_loss: 813.3467
Epoch 7/50
23/23 - 0s - 1ms/step - loss: 949.6710 - val_loss: 663.2787
Epoch 8/50
23/23 - 0s - 1ms/step - loss: 703.8416 - val_loss: 512.1061
Epoch 9/50
23/23 - 0s - 1ms/step - loss: 490.2215 - val_loss: 389.3141
Epoch 10/50
23/23 - 0s - 1ms/step - loss: 359.4137 - val_loss: 312.8605
Epoch 11/50
23/23 - 0s - 1ms/step - loss: 301.4568 - val_loss: 262.0592
Epoch 12/50
23/23 - 0s - 1ms/step - loss: 271.0792 - val_loss: 233.5633
Epoch 13/50
23/23 - 0s - 1ms/step - loss: 253.7307 - val_loss: 213.8286
Epoch 14/50
23/23 - 0s - 1ms/step - loss: 241.1207 - val_loss:

Train and test it 50 times:

In [10]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

# Initialize list of mean squared errors
mse_list = []

# Repeat the process 50 times
for _ in range(50):
    # Split the data
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=None)
    
    # Build the model
    model = regression_model()
    
    # Train the model
    model.fit(X_train, y_train, epochs=50, verbose=0)
    
    # Predict on the test data
    y_pred = model.predict(X_test, verbose=0)
    
    # Compute mean squared error
    mse = mean_squared_error(y_test, y_pred)
    mse_list.append(mse)

print(mse_list)


[118.14382012504592, 139.7038574230891, 102.79368025911762, 141.67150635608232, 124.67995964198654, 113.30873088750786, 133.2430756807141, 152.98247560662082, 149.2900653085746, 143.28911306270606, 120.61916697038396, 136.30859712145454, 147.79695556285677, 133.58760709564032, 109.77134473095754, 144.71526531520968, 132.15287580915972, 117.3092539496814, 145.94227977225523, 127.84897821457517, 133.27648183573746, 90.93791559257676, 136.36622221990746, 136.1265401490851, 122.20011291832095, 143.23044005121784, 150.5771255540155, 124.44154604900793, 140.3592209120131, 126.73555400440793, 147.02739258718563, 140.69059563320835, 122.39073423725445, 147.75239373222018, 114.85629652678085, 95.94510679083123, 141.3650541558098, 139.64868360110586, 137.7616130511026, 122.43932165266587, 146.46645390324673, 148.1971853109503, 125.68692618671585, 151.8430924549709, 132.12978678543394, 104.99025545547823, 142.9108807628334, 155.36797716518754, 128.008367556934, 110.94659446936667]


## Report the mean and standard deviation of the mean squared errors

In [11]:
mean_mse = np.mean(mse_list)
std_mse = np.std(mse_list)
print(f'Mean MSE: {mean_mse}')
print(f'Standard Deviation of MSE: {std_mse}')

Mean MSE: 131.91668960398383
Standard Deviation of MSE: 15.27315166223265


## How does the mean of the mean squared errors compare to that from Step B?

The mean of the mean squared errors (MSE) from the current exercise is lower than the mean MSE from Step B. This indicates that increasing the number of hidden layers from 1 to 3 has further improved the model's performance. Additionally, the standard deviation of the MSE is slightly higher, suggesting a slight increase in variability across different splits of the data.

| Metric | Step A | Step B | Step D |
|--------|--------|--------|------------------|
| Mean MSE | 416.851 | 349.821 | 131.917 |
| Standard Deviation of MSE | 595.123 | 90.653 | 15.273 |




## Conclusion

In this exercise, we have demonstrated the effectiveness of neural networks in predicting numerical values, specifically the compressive strength of concrete. The results indicate that neural networks can serve as powerful predictors when the input data is clean and accurate. By normalizing the data and carefully designing the network architecture, we achieved a significant reduction in the mean squared error (MSE) compared to previous steps.

Furthermore, executing the model multiple times and averaging the results provides a more reliable measure of its accuracy than a single execution. This approach helps to account for variability in the data splits and ensures that the reported performance metrics are robust.

To further improve the model's accuracy and reduce overfitting, two key strategies can be employed: increasing the number of epochs and increasing the number of hidden layers. Increasing the number of epochs allows the model to learn more thoroughly from the training data, while additional hidden layers can capture more complex patterns and interactions within the data. Both measures contribute to enhancing the model's predictive capabilities.

In conclusion, neural networks, when properly configured and trained, can effectively predict numerical outcomes. The iterative process of training and testing, along with architectural adjustments, plays a crucial role in optimizing model performance.